File: src/main/rooms/SettingsRoom.java
package main.rooms;

import main.Slider;
import main.buttons.iconbuttons.Exit;
import main.buttons.iconbuttons.ToggleButton;

import java.io.OutputStream;
import java.io.PrintStream;

public class SettingsRoom extends Room {

    private ToggleButton musicButton;
    private ToggleButton soundButton;
    private ToggleButton groupColour;
    private Exit exit;

    private Slider slider;

    public SettingsRoom() {
        this.musicButton = new ToggleButton("resources/images/music_button.png") {
            @Override
            public void clicked() {
                super.clicked();
                //noinspection StatementWithEmptyBody
                if (this.isToggled()) {
                    main.backgroundMusicThread.interrupt();
                } else {
                    //it will be muted automatically, see SettingsRoom.draw
                }
            }
        };
        this.soundButton = new ToggleButton("resources/images/sound_button.png");
        this.groupColour = new ToggleButton("resources/images/group_colour_button.png");
        this.exit = new Exit();
    }

    @Override
    public void setup() {
        this.musicButton.setToggled(main.getSettings().getBoolean("music"));
        this.soundButton.setToggled(main.getSettings().getBoolean("sound"));
        this.groupColour.setToggled(main.getSettings().getBoolean("group colour"));

        //for hiding the G4P announcement
        PrintStream originalStream = System.out;
        PrintStream dummyStream = new PrintStream(new OutputStream() {
            public void write(int b) {
            }
        });
        System.setOut(dummyStream);
        this.slider = new Slider(main);
        this.slider.setValue(main.getSettings().getFloat("volume"));
        System.setOut(originalStream);
    }

    @Override
    public void draw() {
        this.drawTitle("settings", "settings");

        main.textSize(20);
        main.text(main.getLanguageSelected().getLocalizedString("settings", "volume"), main.screenWidth / 2, 200);

        this.slider.moveTo(main.screenWidth / 2F - this.slider.getWidth() / 2, 220);
        this.slider.draw();

        int length = (ToggleButton.SIZE + ToggleButton.GAP) * 3 - ToggleButton.GAP;
        int x = main.screenWidth / 2 - length / 2;
        this.musicButton.draw(x, 400);
        x += ToggleButton.SIZE + ToggleButton.GAP;
        this.soundButton.draw(x, 400);
        x += ToggleButton.SIZE + ToggleButton.GAP;
        this.groupColour.draw(x, 400);

        this.exit.draw();

        //this needs to be updated immediately because background music constantly plays in the background
        main.backgroundMusic.setGain(main.toGain(this.slider.getValueF()));
        if (this.musicButton.isToggled()) {
            main.backgroundMusic.unmute();
        } else {
            main.backgroundMusic.mute();
        }

        //update settings because they should be in effect immediately
        main.getSettings().put("music", this.musicButton.isToggled());
        main.getSettings().put("sound", this.soundButton.isToggled()); //if this wasn't updated here, the button clicks wouldn't respond to this setting
        main.getSettings().put("volume", this.slider.getValueF());
        main.getSettings().put("group colour", this.groupColour.isToggled());
    }

    @Override
    public void end() {
        this.slider.dispose();
    }

    @Override
    public void mousePressed() {
        this.musicButton.mousePressed();
        this.soundButton.mousePressed();
        this.groupColour.mousePressed();
        this.exit.mousePressed();
    }

}

File: src/main/rooms/SettingsRoom.kt
package main.rooms

import main.Language
import main.Main
import main.Slider
import main.buttons.iconbuttons.Exit
import main.buttons.iconbuttons.IconButton
import main.buttons.iconbuttons.ToggleButton
import java.io.OutputStream
import java.io.PrintStream

object SettingsRoom : Room() {

    private val musicButton: ToggleButton
    private val soundButton: ToggleButton
    private val groupColour: ToggleButton
    private val exit: Exit
    private lateinit var slider: Slider

    init {
        musicButton = object : ToggleButton("resources/images/music_button.png") {
            override fun clicked() {
                super.clicked()
                if (this.isToggled) {
                    Main.backgroundMusicThread.interrupt()
                } else {
                    //it will be muted automatically, see SettingsRoom.draw
                }
            }
        }
        soundButton = ToggleButton("resources/images/sound_button.png")
        groupColour = ToggleButton("resources/images/group_colour_button.png")
        exit = Exit()
    }

    override fun setup() {
        musicButton.isToggled = Main.settings.getBoolean("music")
        soundButton.isToggled = Main.settings.getBoolean("sound")
        groupColour.isToggled = Main.settings.getBoolean("group colour")

        //for hiding the G4P announcement
        val originalStream = System.out
        val dummyStream = PrintStream(object : OutputStream() {
            override fun write(b: Int) {}
        })
        System.setOut(dummyStream)
        slider = Slider(Main)
        slider.setValue(Main.settings.getFloat("volume"))
        System.setOut(originalStream)
    }

    override fun draw() {
        drawTitle("settings", "settings")
        Main.textSize(20f)
        Main.text(Language.languageSelected.getLocalizedString("settings", "volume"), Main.screenWidth / 2f, 200f)
        slider.moveTo(Main.screenWidth / 2f - slider.width / 2, 220f)
        slider.draw()
        val length = (IconButton.SIZE + IconButton.GAP) * 3 - IconButton.GAP
        var x = Main.screenWidth / 2 - length / 2
        musicButton.draw(x.toFloat(), 400f)
        x += IconButton.SIZE + IconButton.GAP
        soundButton.draw(x.toFloat(), 400f)
        x += IconButton.SIZE + IconButton.GAP
        groupColour.draw(x.toFloat(), 400f)
        exit.draw()

        //this needs to be updated immediately because background music constantly plays in the background
        Main.backgroundMusic.gain = Main.toGain(slider.valueF)
        if (musicButton.isToggled) {
            Main.backgroundMusic.unmute()
        } else {
            Main.backgroundMusic.mute()
        }

        //update settings because they should be in effect immediately
        Main.settings.put("music", musicButton.isToggled)
        Main.settings.put("sound", soundButton.isToggled) //if this wasn't updated here, the button clicks wouldn't respond to this setting
        Main.settings.put("volume", slider.valueF)
        Main.settings.put("group colour", groupColour.isToggled)
    }

    override fun end() {
        slider.dispose()
    }

    override fun mousePressed() {
        musicButton.mousePressed()
        soundButton.mousePressed()
        groupColour.mousePressed()
        exit.mousePressed()
    }
}